Add document of Execution Context (#16337)
mdn/content: files/en-us/glossary/Execution Context/index.md

---
title: Execution context
slug: Glossary/Execution_context
tags:
  - Glossary
  - JavaScript
---

> An **execution context** is the environment in which Javascript code executes. It contains the information that links a variable name with the actual variable storing data, either inside the execution context, if they are declared in it, or outside it if not. The binding of the {{jsxref("this")}} reference is also done in the execution context.

Reviewer: ```suggestion
An **execution context** is the environment in which Javascript code executes. It consists of the {{glossary("scope")}}, the information that links a variable name with the actual variable storing data, and of the binding of the {{jsxref("this")}} reference, the object `this` refers to.
```

Reviewer 2: > the information that links a variable name with the actual variable storing data

Can we say "the information that links a variable name to the location storing the data"?

When running, the JavaScript engine initially creates the _global execution context_, then a new context is created for each function or `eval` call.

Variables created using `let` and constants are available in the _current_ execution context only, while variables created using `var` are _hoisted_ to the _global_ execution context and available everywhere, when not hidden by another declaration.

#### Example:
```js
const n = 2;  // n is available to the global execution context

function square(num) { // the square function creates its own execution context, when num is declared
  const ans = num * num; // ans is only available to the function execution context
  return ans;
}

const square2 = square(n); // square2 is available to the global execution context
const square4 = ans; // ans is undefined as it was defined in another execution context

```

## See also

> - {{glossary("Hoisting")}} 

Reviewer: ```suggestion
- {{glossary("Hoisting")}} 
- {{glossary("Scope")}}
```

---

> title: Execution Context

Reviewer: ```suggestion
title: Execution context
```

> slug: Glossary/Execution Context

Reviewer: ```suggestion
slug: Glossary/Execution_context
```

tags:
  - Glossary
  - JavaScript
---

> Everything in JavaScript happens/executes inside **Execution Context**.

Reviewer: We want a _definition_  of an execution context here.
```suggestion
An **execution context** is the environment in which Javascript code executes. It contains the information that links a variable name with the actual variable storing data, either inside the execution context, if they are declared in it, or outside it if not. The binding of the {{jsxref("this")}} reference is also done in the execution context.

When running, the JavaScript engine initially creates the _global execution context_, then a new context is created for each function or `eval` call.

Variables created using `let` and constants are available in the _current_ execution context only, while variables created using `var` are _hoisted_ to the _global_ execution context and available everywhere, when not hidden by another declaration.
```

#### Example:
```js

> var n = 2;

Reviewer 2: ```suggestion
const n = 2;
```

Author: I am using let in place of const. Is it valid?

Reviewer 2: If you are not assigning any value to a variable after the declaration then keep it `const`.
[Refer the guidelines](https://developer.mozilla.org/en-US/docs/MDN/Guidelines/Code_guidelines/JavaScript#declaring_variables)

Author: Thank you, Since I was unable to apply your changes. I make it by myself. Please check.

Author: @OnkarRuikar Can you please approve the same if it seems to be fine?

Reviewer 2: @ayushikolay final review and merge will be done by code owners: @dipikabh or @teoli2003.

function square(num) {

>   var ans = num * num;

Reviewer 2: ```suggestion
  const ans = num * num;
```

  return ans;
}

var square2 = square(n);

> var square4 = square(4);

Reviewer 2: ```suggestion
const square4 = square(4);
```

> const n = 2;
> 
> function square(num) {
>   const ans = num * num;
>   return ans;
> }
> 
> const square2 = square(n);
> const square4 = square(4);
> 

Reviewer: ```suggestion
const n = 2;  // n is available to the global execution context

function square(num) { // the square function creates its own execution context, when num is declared
  const ans = num * num; // ans is only available to the function execution context
  return ans;
}

const square2 = square(n); // square2 is available to the global execution context
const square4 = ans; // ans is undefined as it was defined in another execution context

```

```

> ### Inside the Execution Context there two components:
> 
> #### 1. Memory component/ Variable Environment
> This is the first phase also known as `Memory Creation Phase` in which, JavaScript will allocate memory to all its variables and functions.
> 
> Variables and functions values can be stored in a  key-value format. Memory component is also called as `variable environment`.
> 
> 
> | Memory Component                                  | Code Component    | 
> | --------------------------------------------------| :----------------:| 
> | n : undefined                                     |                   | 
> | square: f() { var ans = num* num; return ans}     |                   |  
> | square2: undefined                                |                   |
> | square4: undefined                                |                   |
> 
> For variable(s), key is the variable name itself and value is undefined (even if the variable is initialized).
> For function(s), key is the function name and value is body of the code.
> 
> #### 2. Code component/ Thread Of Execution
> This is the second phase also known as `Code Execution Phase`, where JavaScript will run again line by line and it starts executing the code now.
> 
> It starts assigning the values to the variables in memory.
> 
> Code Component is also called as `Thread of Execution`. 
> 
> #### When the code is run `Global Execution Context` is created.
> 
> #### It is due to the **Execution Context**, the variable(s) name and function(s) are hoisted because of allocation in memory.
> 

Reviewer: I don't think this brings much in the understanding of an execution context.

```suggestion
```

> ### Two golden Rules of Hoisting:
> So, the concept of Hoisting can be explained now based on the Execution Context.
> 
> 1. Variable(s) are scanned and are made undefined(Memory Creation Phase).
> 2. Function(s) are scanned and are made available(Memory Creation Phase).

Reviewer: Let's keep hoisting in its glossary entry, and keep the entry here for execution context:
```suggestion

## See also

- {{glossary("Hoisting")}} 
```
